app/src/androidTest/java/org/mozilla/fenix/helpers/RetryTestRule.kt

```diff
@@ -23,30 +23,51 @@ class RetryTestRule(private val retryCount: Int = 5) : TestRule {
                     base.evaluate()
                     break
                 } catch (t: AssertionError) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: AssertionFailedError) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: UiObjectNotFoundException) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: NoMatchingViewException) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: IdlingResourceTimeoutException) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: RuntimeException) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
                     if (i == retryCount) {
                         throw t
                     }
                 } catch (t: NullPointerException) {
+                    for (resource in IdlingRegistry.getInstance().resources) {
+                        IdlingRegistry.getInstance().unregister(resource)
+                    }
```

Comment:
For the above unregister calls, I'm not opposed to it but I'm not sure if this a good place. I'd prefer to keep them closer to the register calls in test. Do we know if we could place these preferably in the `tearDown` functions instead? Or write one function that checks for instances as you did above and and unregisters.

Reply:
> I'd prefer to keep them closer to the register calls in test. Do we know if we could place these preferably in the `tearDown` functions instead?

Well, that's the problem: if the test fails before it can reach the unregister calls, it will restart the app and then it crashes. The tearDown methods aren't called until after the 3 retries are done. So we have to make sure to keep unregistering after the registration as before, and also to do this while in the RetryTestRule loop for exceptions like this.

Reply:
move this code to the IdlingResourceHelper.kt class